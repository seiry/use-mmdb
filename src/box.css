.box {
  background-color: rgb(var(--background-end-rgb));
  border-radius: 10px;
  padding: 10px;
  margin: 10px;
  box-shadow: 0 0 10px rgba(var(--background-start-rgb), 0.5);
  width: 30rem;
}
