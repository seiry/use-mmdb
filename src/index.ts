export * from './hook';
export * from './webWorker';
export * from './getWoker';
