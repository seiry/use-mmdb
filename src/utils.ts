export async function fetchFileAsBuffer(fileUrl: string) {
  try {
    const response = await fetch(fileUrl);
    if (!response.ok) {
      throw new Error('Network response was not ok');
    }
    const arrayBuffer = await response.arrayBuffer();
    // 例如，创建一个 Uint8Array 视图
    const uint8View = new Uint8Array(arrayBuffer);
    const buf = Buffer.from(uint8View);
    return buf;
  } catch (error) {
    console.error('Fetching file failed:', error);
  }
}
