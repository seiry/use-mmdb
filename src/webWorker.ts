import { expose } from 'comlink';

const obj = {
  counter: 12,
  inc() {
    this.counter++;
  },
};

expose(obj);

export type Obj = typeof obj;
