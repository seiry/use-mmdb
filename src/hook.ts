import { useEffect, useState } from 'react';
import * as mmdb from 'mmdb.js';
import { fetchFileAsBuffer } from './utils';

const _mmdbUrl =
  'https://raw.githubusercontent.com/Max-Sum/17mon-mmdb/release/Country.mmdb';

export const useIp = (mmdbUrl?: string) => {
  const [reader, setReader] = useState<mmdb.Reader<mmdb.CityResponse>>();
  useEffect(() => {
    fetchFileAsBuffer(mmdbUrl ?? _mmdbUrl).then((res) => {
      if (!res) return;
      const reader = new mmdb.Reader<mmdb.CityResponse>(res);
      setReader(reader);
    });
  }, [mmdbUrl]);

  return reader;
};
