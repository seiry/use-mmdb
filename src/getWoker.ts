import { wrap } from 'comlink';
import { Obj } from './webWorker';

const worker = new Worker(new URL('./webWorker.ts', import.meta.url));
export const workerObj = wrap<Obj>(worker);
