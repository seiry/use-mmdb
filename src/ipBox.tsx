import { FC } from 'react';
import { useIp } from './hook';
import { prettyPrintJson } from 'pretty-print-json';
import './box.css';

export interface IpBoxProps {
  ip: string;
}

export const IpBox: FC<IpBoxProps> = (props) => {
  const { ip } = props;
  const reader = useIp();
  const data = reader?.get(ip);
  const html = prettyPrintJson.toHtml(data);

  return (
    <div className="box ">
      ip address: <pre>{ip}</pre>
      <pre dangerouslySetInnerHTML={{ __html: html }}></pre>
    </div>
  );
};
