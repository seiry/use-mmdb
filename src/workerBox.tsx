import { wrap } from 'comlink';
import { Obj } from './webWorker';
import { IpBoxProps } from './ipBox';
import { prettyPrintJson } from 'pretty-print-json';
import { FC, useState, useEffect } from 'react';

const worker = new Worker(new URL('./webWorker.ts', import.meta.url));
const obj = wrap<Obj>(worker);

export const IpBoxWebworker: FC<IpBoxProps> = (props) => {
  const { ip } = props;
  const [html, setHtml] = useState('');
  useEffect(() => {
    obj.getIPInfo(ip).then((res) => {
      setHtml(prettyPrintJson.toHtml(res));
    });
  }, [ip]);

  return (
    <div className="box ">
      ip address: <pre>{ip}</pre>
      <pre dangerouslySetInnerHTML={{ __html: html }}></pre>
    </div>
  );
};
