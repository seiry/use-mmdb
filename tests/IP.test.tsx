import React from 'react';
import renderer from 'react-test-renderer';
import { expect, it } from 'vitest';
import { IpBox } from '../src/ipBox';

it('renders 1.1.1. correctly', () => {
  const tree = renderer.create(<IpBox ip="1.1.1.1" />).toJSON();
  expect(tree).toMatchSnapshot();
});
