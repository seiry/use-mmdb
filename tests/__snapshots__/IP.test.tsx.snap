// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders 1.1.1. correctly 1`] = `
<div
  className="box "
>
  ip address: 
  <pre>
    1.1.1.1
  </pre>
  <pre
    dangerouslySetInnerHTML={
      {
        "__html": "<span class=json-number>undefined</span>",
      }
    }
  />
</div>
`;
